### PR TITLE
feature: Noise models

### PR DESCRIPTION
**Pull request by speller26**
Introduces `NoiseModel`, which encapsulates a collection of noise channels to be applied to a given circuit according to some criteria.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-braket-sdk-python/blob/main/README.md) and [API docs](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by speller26** [comment]
Note: parametrized noise models will be developed in #316 instead of this PR.

File: src/braket/circuits/noise.py [hunk 1]
```diff
@@ -128,92 +145,129 @@ def probability(self) -> float:
     def __repr__(self):
         return f"{self.name}('probability': {self.probability}, 'qubit_count': {self.qubit_count})"
 
+    def __str__(self):
+        return f"{self.name}({self.probability})"
+
+    @property
+    def parameters(self) -> List[Union[FreeParameter, float]]:
+        """
+        Returns the free parameters associated with the object.
```
[the remark this answers is not in the document]
Comment (speller26) [reply src/braket/circuits/noise.py:154]: Could be either, as is the case with [AngledGate](https://github.com/aws/amazon-braket-sdk-python/blob/main/src/braket/circuits/angled_gate.py#L58)

File: src/braket/circuits/noise.py [hunk 2]
```diff
@@ -128,92 +145,129 @@ def probability(self) -> float:
...
+            into this object using the `from_dict()` method.
+        """
+        return {
+            "__class__": self.__class__.__name__,
+            "probability": self.probability,
+            "qubit_count": self.qubit_count,
+            "ascii_symbols": self.ascii_symbols,
+        }
+
+    def bind_values(self, **kwargs):
```
[the remark this answers is not in the document]
Comment (speller26) [reply src/braket/circuits/noise.py:182]: Our convention for gates is to return a new object, so the original can be reused.

File: src/braket/circuits/noise.py [hunk 3]
```diff
@@ -128,92 +145,129 @@ def probability(self) -> float:
...
+        """
+        Takes in parameters and attempts to assign them to values.
```
[the remark this answers is not in the document]
Comment (speller26) [reply src/braket/circuits/noise.py:184]: This is the convention for gates, and the reason for this is the ability to pass the same parameter map dict to multiple different parametrized objects.

File: src/braket/circuits/noise.py [hunk 4]
```diff
@@ -470,3 +713,67 @@ def probability(self) -> float:
...
+        """
+        return [self.gamma, self.probability]
+
+    def __eq__(self, other):
+        if isinstance(other, type(self)):
+            return (
+                self.name == other.name
+                and self.gamma == other.gamma
+                and self.probability == other.probability
+            )
```
[the remark this answers is not in the document]
Comment (speller26) [reply src/braket/circuits/noise.py:737]: We currently don't do this for gates, so this is just for consistency. We can make sweeping changes in another PR.

File: src/braket/circuits/noise.py [hunk 5]
```diff
@@ -301,27 +343,107 @@ def __init__(
...
+    @property
+    def parameters(self) -> List[Union[FreeParameter, float]]:
+        """
+        Returns the free parameters associated with the object.
+
+        Returns:
+            Union[FreeParameter, float]: Returns the free parameters or fixed value
+            associated with the object.
+        """
+        return list(self._probabilities.values())
```
[the remark this answers is not in the document]
Comment (speller26) [reply src/braket/circuits/noise.py:385]: I'll re-add the `probabilities` property and document that the parameters in in alphabetical order of the supplied Pauli strings.

File: src/braket/circuits/noise.py [hunk 6]
```diff
@@ -331,85 +453,152 @@ def __init__(
...
+    @property
+    def parameters(self) -> List[Union[FreeParameter, float]]:
+        """
+        Returns the free parameters associated with the object.
+
+        Returns:
+            Union[FreeParameter, float]: Returns the free parameters or fixed value
+            associated with the object.
+        """
+        return self._parameters
```
[the remark this answers is not in the document]
Comment (speller26) [reply src/braket/circuits/noise.py:526]: Added order in docs

File: src/braket/circuits/noise.py [hunk 7]
```diff
@@ -470,3 +713,67 @@ def probability(self) -> float:
...
+    @property
+    def parameters(self) -> List[Union[FreeParameter, float]]:
+        """
+        Returns the free parameters associated with the object.
+
+        Returns:
+            Union[FreeParameter, float]: Returns the free parameters or fixed value
+            associated with the object.
+        """
+        return [self.gamma, self.probability]
```
[the remark this answers is not in the document]
Comment (speller26) [reply src/braket/circuits/noise.py:729]: Added order in docs

File: src/braket/circuits/noise_model/noise_model.py [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+        if not readout_noise_instructions:
+            return circuit
+
+        result_types = circuit.result_types
+        if result_types:
+            for result_type in result_types:
+                target_qubits = list(result_type.target)
+                for item in readout_noise_instructions:
+                    if item.criteria.result_type_matches(result_type):
+                        circuit.apply_readout_noise(noise=item.noise, target_qubits=target_qubits)
```
[the remark this answers is not in the document]
Comment (kshitijc) [reply src/braket/circuits/noise_model/noise_model.py:349]: Summoning @mbeach-aws and @licedric for help here.

File: src/braket/circuits/noise.py [hunk 9]
```diff
@@ -72,11 +74,19 @@ def to_matrix(self, *args, **kwargs) -> Any:
...
         if isinstance(other, Noise):
             return self.name == other.name
-        return NotImplemented
+        return False
 
     def __repr__(self):
         return f"{self.name}('qubit_count': {self.qubit_count})"
 
+    @classmethod
+    def from_dict(cls, noise: dict) -> "Noise":
```
[the remark this answers is not in the document]
Comment (krneta) [reply src/braket/circuits/noise.py:83]: Will do.

File: src/braket/circuits/noise.py [hunk 10]
```diff
@@ -331,85 +453,152 @@ def __init__(
...
 
+        Returns:
+            dict: A dictionary object that represents this object. It can be converted back
+            into this object using the `from_dict()` method.
+        """
+        return {
+            "__class__": self.__class__.__name__,
+            "probX": self.probX,
+            "probY": self.probY,
+            "probZ": self.probZ,
```
[the remark this answers is not in the document]
Comment (krneta) [reply src/braket/circuits/noise.py:540]: I wanted some way to be able to serialize and deserialize. I thought I would need to keep track of the type (class) of each object and the values, and recursively call all of the children. 

If I can do that with __dict__, I would be happy to. But it seems like I would just need to implement __dict__ (rather than  to_dict()) for each class instead. Would that be better?

File: src/braket/circuits/__init__.py [hunk 11]
```diff
@@ -30,6 +30,20 @@
 from braket.circuits.instruction import Instruction  # noqa: F401
 from braket.circuits.moments import Moments, MomentsKey  # noqa: F401
 from braket.circuits.noise import Noise  # noqa: F401
+from braket.circuits.noise_model import (  # noqa: F401
```
[the remark this answers is not in the document]
Comment (krneta) [reply src/braket/circuits/__init__.py:33]: OK, are there specific names you are worried about? Criteria and CriteriaKey seems similar to Moments and MomentsKey, to me, and it's nice not to have to know that Noise and NoiseModel need to be imported from different modules.

File: src/braket/circuits/noises.py [hunk 12]
```diff
@@ -111,6 +112,38 @@ def bit_flip(target: QubitSetInput, probability: float) -> Iterable[Instruction]
...
+        """
+        Converts a dictionary representation of this class into this class.
+
+        Args:
+            noise(dict): The dictionary representation of this noise.
+
+        Returns:
+            Noise: A Noise object that represents the passed in dictionary.
+        """
+        return BitFlip(probability=noise["probability"])
```
[the remark this answers is not in the document]
Comment (krneta) [reply src/braket/circuits/noises.py:145]: I'm not sure. Is that a problem? I could verify the class name in the dictionary, but it seems unnecessary. Why not allow customers to do this, if they want to?

File: src/braket/circuits/noise.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (krneta) [reply src/braket/circuits/noise.py:155]: Will update docs.

File: src/braket/circuits/__init__.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (krneta) [reply src/braket/circuits/__init__.py:33]: OK, will move out to noise_model

File: src/braket/circuits/noises.py [hunk 13]
```diff
@@ -1320,11 +1330,7 @@ def kraus(
         )
 
     def to_dict(self) -> dict:
-        return {
-            "__class__": self.__class__.__name__,
-            "matrices": self._matrices,
```
Comment (kshitijc) [inline src/braket/circuits/noises.py]: Could we serialize this as `self._matrices.tolist()` ?

Comment (krneta) [reply src/braket/circuits/noises.py]: Yes, I was hoping to avoid it in this PR, because from the doc:
"The array may be recreated via a = np.array(a.tolist()), although this may sometimes lose precision."

Comment (krneta) [reply src/braket/circuits/noises.py]: Do you think that would be OK?

Comment (krneta) [reply src/braket/circuits/noises.py]: Also, it contains "complex" objects, which are not serializable.

Comment (kshitijc) [reply src/braket/circuits/noises.py]: Oh, ok. In that case we can leave it unimplemented for now.

File: src/braket/circuits/noise_model/noise_model.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (krneta) [reply src/braket/circuits/noise_model/noise_model.py]: @mbeach-aws Can you take a look at the updated [tests](https://github.com/aws/amazon-braket-sdk-python/pull/313/files#diff-925e67ef63651dabc9c31461b9ef409893e711a72b867e2dfeed2bcc7ff5b7eeR428) for the noise model? It may not open up automatically because the diff is large. The unit tests show how I now handle multiple result types and when I use the default.

[the remark this answers is not in the document]
Comment (krneta) [reply src/braket/circuits/noise_model/noise_model.py]: I've now updated the code to not have any assumptions about what is being measured when there are no default result types in the circuit. This means that the customers will need to add observables to the circuit manually, if they want to have the noise model apply readout noise.

File: test/unit_tests/braket/circuits/noise/test_observable_criteria.py [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+            [ResultType.Sample(Observable.X(), 4)],
+        ),
+        (
+            None,
+            None,
+            [
+                ResultType.Sample(Observable.X(), 0),
+                ResultType.Sample(Observable.Z(), 10),
+            ],
+            [ResultType.Probability(0)],
```
Comment (mbeach-aws) [inline test/unit_tests/braket/circuits/noise/test_observable_criteria.py:92]: jsut double checking that it doesn't match on probability result types.

File: test/unit_tests/braket/circuits/noise/test_gate_criteria.py [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+            [Instruction(Observable.X, 0)],
+        ),
+        (
+            Gate.H,
+            range(3),
+            [Instruction(Gate.H(), 0), Instruction(Gate.H(), 1)],
+            [
+                Instruction(Gate.H(), 3),
+                Instruction(Gate.I(), 1),
+                [Instruction(Gate.H(), 0)],
```
Comment (mbeach-aws) [inline test/unit_tests/braket/circuits/noise/test_gate_criteria.py:95]: same as unitary_gate_test.py file,  why is this final element a list?

File: test/unit_tests/braket/circuits/noise/test_unitary_gate_criteria.py [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+@pytest.mark.parametrize(
+    "unitary, qubits",
+    [
+        (None, 1),
+        ([], 1),
+        ([h_unitary()], 1),
+        (np.zeros(15, dtype=int), 1),
+    ],
+)
+def test_invalid_params(unitary, qubits):
```
[the remark this answers is not in the document]
Comment (krneta) [reply test/unit_tests/braket/circuits/noise/test_unitary_gate_criteria.py:143]: Will do!

File: test/unit_tests/braket/circuits/noise/test_unitary_gate_criteria.py [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+            None,
+            [
+                Instruction(h_unitary(), 0),
+                Instruction(h_unitary(), 1),
+            ],
+            [
+                Instruction(Gate.H, 3),
+                Instruction(h_unitary(), 3),
+                Instruction(i_unitary(), 1),
+                [Instruction(h_unitary(), 0)],
```
[the remark this answers is not in the document]
Comment (krneta) [reply test/unit_tests/braket/circuits/noise/test_unitary_gate_criteria.py:89]: Again, this was a leftover from when we allowed an iterable for the matchers (to support layers). We've removed this as a requirement, but I didn't remove the test. Right now it will not match any such instructions, which I think is correct, but we could also throw an exception, if you prefer.

File: test/unit_tests/braket/circuits/noise/test_unitary_gate_criteria.py [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+@pytest.mark.parametrize(
+    "unitary, qubits, defaults, matching_instructions, non_matching_instructions",
+    [
+        (
+            h_unitary(),
+            range(3),
+            None,
+            [
+                Instruction(h_unitary(), 0),
+                Instruction(h_unitary(), 1),
```
[the remark this answers is not in the document]
Comment (krneta) [reply test/unit_tests/braket/circuits/noise/test_unitary_gate_criteria.py:83]: 0 covers the edge case (first number we support), 1 covers the regular cases (a non-edge case number we support). We can add 2, and other numbers, but they're just more regular cases, so they shouldn't be needed - unless I'm missing something?

File: test/unit_tests/braket/circuits/noise/test_unitary_gate_criteria.py [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+    for instruction in matching_instructions:
+        assert criteria.instruction_matches(instruction)
+    for instruction in non_matching_instructions:
+        assert not criteria.instruction_matches(instruction)
+
+
+@pytest.mark.parametrize(
+    "unitary, qubits",
+    [
+        (None, 1),
```
[the remark this answers is not in the document]
Comment (krneta) [reply test/unit_tests/braket/circuits/noise/test_unitary_gate_criteria.py:137]: They're all the same error - the passed parameter is not the type that's expected (ie. an np array is not a valid type either). We can make them all type errors, if you think that's better.

File: test/unit_tests/braket/circuits/noise/test_unitary_gate_criteria.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (mbeach-aws) [reply test/unit_tests/braket/circuits/noise/test_unitary_gate_criteria.py:83]: right, I just figured since we're doing `range(3)`, we should just test it matches all three instructions: ` Instruction(h_unitary(), i)` for `i = 0,1,2` to be consistent. Otherwise why not just use `range(2)`?

File: test/unit_tests/braket/circuits/noise/test_unitary_gate_criteria.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (mbeach-aws) [reply test/unit_tests/braket/circuits/noise/test_unitary_gate_criteria.py:137]: yes, you're right. I forgot that the numpy array must be wrapped with Unitary.

File: test/unit_tests/braket/circuits/noise/test_observable_criteria.py (hunk 14 above)
Comment (krneta) [reply test/unit_tests/braket/circuits/noise/test_observable_criteria.py:92]: Yes, that's exactly what the test does.

File: src/braket/circuits/noise_model/noise_model.py [hunk 20]
```diff
@@ -0,0 +1,409 @@
...
+
+        Args:
+            circuit (Circuit): A circuit to apply `noise` to.
+            readout_noise_instructions (List[NoiseModelInstruction]): The list of readout noise
+                to apply.
+
+        Returns:
+            Circuit: The passed in circuit, with the readout noise applied.
+        """
+        if not readout_noise_instructions or not circuit.result_types:
```
Comment (kshitijc) [inline src/braket/circuits/noise_model/noise_model.py:340]: If the customer doesn't specify any result types, the default behavior is to measure all qubits. In such a case, how should the customers model this readout noise ?

File: src/braket/circuits/noise_model/noise_model.py [hunk 21]
```diff
@@ -0,0 +1,409 @@
...
+
+        Args:
+            noise (Noise): The noise to add.
+            criteria (Criteria): The criteria that determines when the noise will be applied.
+
+        Returns:
+            NoiseModel: This NoiseModel object.
+
+        Raises:
+            ValueError: If the noise type is not specified and can not be inferred.
```
[the remark this answers is not in the document]
Comment (krneta) [reply src/braket/circuits/noise_model/noise_model.py:132]: Good catch! This is doc is actually wrong. It's a leftover from when the noise parameter is optional. Removed it.

File: src/braket/circuits/noise_model/noise_model.py (hunk 20 above)
Comment (krneta) [reply src/braket/circuits/noise_model/noise_model.py:340]: After a few iterations - we had issues dealing with edge cases with different types of result types - we decided that customers will need to add a specific result type if they wanted to add noise to it.

**Review by kshitijc — DISMISSED** [review]

**Review by mbeach-aws — APPROVED** [review]
